Use zrecompile, create digest file, improve clean cache, don't zcompile zcompdump in completion, move templates/zlogin to login_init.zsh and zmanage to init.zsh

ericbn opened the pull request:
This wraps up the suggestions and discussions of PRs #216 and #217

ericbn commented:
@Eriner, I've separated this in commits that do separate changes. Maybe reviewing commits individually will help better going through all that is going on...

Eriner commented:
A cursory look over the changes looks good. I'll pull this into a testing VM and give it more time tomorrow evening. I'm particularly interested in benchmarking this startup performance in the case that I intentionally throttle the disk performance to simulate less-capable hardware.

Eriner commented:
Great work @ericbn! :+1: :)

ericbn commented:
@Eriner, I've been using this code for almost a month without issues. There's one change that I think should be made though: 

The `modules/meta/functions/zmanage` function in zcompiled into `${ZIM_HOME}/functions.zwc` (which in turn is added to the `fpath`). The problem is that if the `functions.zwc` gets corrupted, then `zmanage` won't be available, and it is needed exactly to clear the cache, which would solve the issue (so the average user would get into a kind of deadlock).

My suggestion is that we stop having a `meta´ module, and move the `zmanage` function declaration to maybe inside the `init.zsh` file. What do you think?

Eriner commented:
@ericbn Yes, I too have been running this on a VM without issue. :+1: to the zmanage comment; I think that's a good idea.

ericbn commented:
@Eriner, cool, just pushed a commit with the zmanage changes.

partounian commented:
What else needs to be done here?

ericbn commented:
@Eriner, maybe the last thing we need here starting with this change and going forward are Release Notes / a Change Log.

As we don't have version numbers, maybe it could be grouped by date. For the current changes it could be something like:

```
## 2017-11-23

Major changes in the core code. Using a functions.zwc digest file for all
zcompiled module functions (fixes #86). Changed the template files, and removed
the meta module.

The following actions are required after the update:

* In `.zimrc`: remove the meta module from zmodules
* In `.zlogin`: replace previous code by

      [[ -s ${ZIM_HOME}/login_init.zsh ]] && source ${ZIM_HOME}/login_init.zsh

* It is advisable to also clean the cache with

      $ zmanage clean-cache
```

@Eriner, what do you think?

Eriner commented:
@ericbn I will dedicate time tomorrow evening to finally making the move to the zimframework repo. I think the best bet is to just xfer this repo to /core, and then we can make the rest of the changes as necessary. Once that full xfer is done, we'll release 1.0 and start tagging versions. Changelogs like that will be part of that.

As for the actual commits here, I don't have any issues with these commits. Feel free to merge. Good stuff! :+1:

ericbn commented:
@Eriner, now that I understand the initialization process better, after working on this, I think I'm better prepared to contribute on the planned changes on splitting the modules in separate repos. I even think the initialization code (mostly everything that was touched in this PR) should not change because of these next changes, and they could be handled as independent from each other.

Maybe we can arrange a pair programming session to look at the next changes together.

Also, I've been opening some PRs with other changes for your comments on them.

Eriner commented:
@ericbn sure, are you available on Sunday? I have some preliminary docs for planned changes written, but that's about as far as I ended up getting. Also, that is on a different machine that I won't have access to until the weekend; I've been traveling for work. If Sunday works for you, I'll drop a video chat link in here for us and anyone else who is interested and might want to give feedback.

ericbn commented:
@Eriner, this Sunday works for me. Good idea!

I'm at GMT-5, so let's say anytime between 10AM-10PM GMT-5 is good.

Eriner commented:
@ericbn Okay, we'll come to a more definite time as Sunday approaches.

Eriner commented:
@ericbn Let's plan on starting in an hour, at noon EST: https://meet.jit.si/IndifferentDwarfsEatCuriously

I'll likely only be available for around an hour to 90 minutes, but we'll blitz through what we can.

partounian commented:
I set a reminder on my phone for this but it didn't go off. How's it going?
Does it look like I could hop on and help with something?

On Sun, Dec 3, 2017, 7:57 AM Matt Hamilton <notifications@github.com> wrote:

> @ericbn <https://github.com/ericbn> Let's plan on starting in an hour, at
> noon EST: https://meet.jit.si/IndifferentDwarfsEatCuriously
>
> I'll likely only be available for around an hour to 90 minutes, but we'll
> blitz through what we can.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Eriner/zim/pull/218#issuecomment-348787624>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACbIGyRvKW45wR8HvKjx3bx7xVnqbL0Vks5s8sTQgaJpZM4PoFsL>
> .
>
-- 

Thank you,
Patrick Artounian
